File: SolutionCodes/Q17.cpp
//240. Search a 2D Matrix II
//160 ms solution

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = matrix[0].size();
        int i = 0;
        int j = cols-1;

        while(i<rows and j >=0)
        {   int news = matrix[i][j];
            if(news>target)
                j--;
            else if(news< target)
                i++;

            if(news==target)
                return true;
        }
        return false;
    }
};


//sample 64 ms submission
/*
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = 0;
        int col = matrix[0].size() - 1;

        bool found = false;
        while(row < matrix.size() && col >= 0)
        {
            int currentVal = matrix[row][col];
            if (currentVal == target)
            {
                found = true;
                break;
            }

            if (target > currentVal)
                row++;
            else
                col--;
        }
        return found;
    }
};
*/

File: SolutionCodes/Q17.py
# 240. Search a 2D Matrix II

# 250 ms
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])
        i = 0
        j = cols - 1
        while i < rows and j >= 0:
            if matrix[i][j] == target:
                return True
            if matrix[i][j] < target:
                i += 1
            else:
                j -= 1
        return False


# # sample 140 ms submission
# class Solution:
#     def searchMatrix(self, matrix, target):
#         """
#         :type matrix: List[List[int]]
#         :type target: int
#         :rtype: bool
#         """
#         if len(matrix) == 0 or len(matrix[0]) == 0:
#             return False
#
#         row = -1
#
#         for i in range(len(matrix)):
#             if matrix[i][0] == target:
#                 return True
#             if matrix[i][0] > target:
#                 row = i
#                 break
#         if row == -1:
#             row = len(matrix) - 1
#
#         col = 1
#         while row >= 0 and col < len(matrix[row]):
#             if matrix[row][col] == target:
#                 return True
#             if matrix[row][col] > target:
#                 row -= 1
#             else:
#                 col += 1
#         return False
